src/main/java/org/testng/xml/XmlSuite.java

```diff
@@ -1017,5 +1019,114 @@ public void addTest(XmlTest test) {
+   */
+  public XmlSuite cloneIfContainsTestsWithNamesMatchingAny(List<String> testNames) {
+    if (testNames == null || testNames.isEmpty()) {
+      throw new TestNGException("Please provide a valid list of names to check.");
+    }
+
+    //Search through all the child suites.
+    for (XmlSuite suite : getChildSuites()) {
+      XmlSuite clonedSuite = cloneIfSuiteContainTestsWithNamesMatchingAny(suite, testNames);
+      if (clonedSuite != null) {
+        return clonedSuite;
+      }
+    }
+
+    //Tests weren't found in child suites. Lets search in the current suite.
+    XmlSuite clonedSuite = cloneIfSuiteContainTestsWithNamesMatchingAny(this, testNames);
+
+    if (clonedSuite == null) {
+      throw new TestNGException("The test(s) <" + testNames.toString() + "> cannot be found.");
+    }
+    return clonedSuite;
+  }
+
+  private XmlSuite cloneIfSuiteContainTestsWithNamesMatchingAny(XmlSuite suite, List<String> testNames) {
+    List<XmlTest> tests = Lists.newLinkedList();
+    for (XmlTest xt : suite.getTests()) {
+      if (xt.nameMatchesAny(testNames)) {
+        tests.add(xt);
+      }
+    }
+    if (tests.isEmpty()) {
+      return null;
+    }
+    return cleanClone(tests);
+  }
+
+  /**
+   * A validator that runs through the list of suites and checks if each of the suites contains
+   * any {@link XmlTest} with the same name. If found, then a {@link TestNGException} is raised.
+   *
+   * @param suites - The list of {@link XmlSuite} to validate.
+   */
+  public static void validateIfSuitesContainDuplicateTests(List<XmlSuite> suites) {
+    for (XmlSuite suite : suites) {
+      suite.ensureNoDuplicateTestsArePresent();
+      validateIfSuitesContainDuplicateTests(suite.getChildSuites());
+    }
+  }
+
+  /**
+   * Ensures that the current suite doesn't contain any duplicate {@link XmlTest} instances.
+   * If duplicates are found, then a {@link TestNGException} is raised.
+   */
+  public void ensureNoDuplicateTestsArePresent() {
+    Set<String> testNames = Sets.newHashSet();
+    for (XmlTest test : getTests()) {
+      if (!testNames.add(test.getName())) {
+        throw new TestNGException("Two tests in the same suite [" + getName() + "] "
+                + "cannot have the same name: " + test.getName());
+      }
+    }
+  }
+
+  /**
+   * Ensure that two XmlSuite don't have the same name
+   * @param suites - The List of {@link XmlSuite} that are to be tested and names updated if duplicate
+   *               names found.
+   */
+  public static void adjustSuiteNamesToEnsureUniqueness(List<XmlSuite> suites) {
+    adjustSuiteNamesToEnsureUniqueness(suites, Sets.<String>newHashSet());
+  }
+
+  private static void adjustSuiteNamesToEnsureUniqueness(List<XmlSuite> suites, Set<String> names) {
+    for (XmlSuite suite : suites) {
+      String name = suite.getName();
+
+      int count = 0;
+      String tmpName = name;
```

Comment:
Why do we need to use `StringBuilder` here ? We are not building the string in the loop. We are just overwriting its value with the previous value. A simple concatenation of strings is literally equal to what the StringBuilder does especially when the concatenation is not being done to construct a string in a loop.

Reply:
StringBuilder possible here?